Index: src/benchmark/testdriver/SPARQLConnection.java
===================================================================
--- src/benchmark/testdriver/SPARQLConnection.java	(revision 77)
+++ src/benchmark/testdriver/SPARQLConnection.java	(working copy)
@@ -107,6 +107,7 @@
 		int queryMixRun = queryMix.getRun() + 1;
 
 		InputStream is = qe.exec();
+		System.out.println("(" + queryMixRun + "," + queryNr + "," + queryType + ") running...");
 
 		if(is==null) {//then Timeout!
 			double t = this.timeout/1000.0;
@@ -142,6 +143,7 @@
 	                   resultCount);
 		
 		queryMix.setCurrent(resultCount, timeInSeconds);
+		System.out.println("(" + queryMixRun + "," + queryNr + "," + queryType + ") done.");
 		qe.close();
 	}
 	

